server/src/main/java/org/opensearch/action/ActionModule.java

```diff
@@ -1072,4 +1100,81 @@ public void unregisterDynamicAction(ActionType<?> action) {
             return registry.get(action);
         }
     }
+
+    /**
+     * The DynamicRouteRegistry maintains a registry mapping {@link org.opensearch.rest.RestHandler.Route} instances to {@link org.opensearch.rest.extensions.RestSendToExtensionAction} instances.
+     * <p>
+     * This class is modeled after {@link NamedRegistry} but provides both register and unregister capabilities.
+     *
+     * @opensearch.internal
+     */
+    public static class DynamicRouteRegistry {
+        // This is an instance of a DynamicActionRegistry containing all registered transport actions
+        private DynamicActionRegistry dynamicActionRegistry;
+        // A dynamic registry to add or remove Route / RestSendToExtensionAction pairs
+        // at times other than node bootstrap.
+        private final Map<RestHandler.Route, RestSendToExtensionAction> registry = new ConcurrentHashMap<>();
+
+        private final Set<String> registeredRestActionNames = new HashSet<>();
+
+        /**
+         * Register the immutable actions in the registry.
+         *
+         * @param dynamicActionRegistry A registry of all transport actions - native and dynamic
+         */
+        public void registerDynamicActionRegistry(DynamicActionRegistry dynamicActionRegistry) {
+            this.dynamicActionRegistry = dynamicActionRegistry;
+        }
+
+        /**
+         * Add a dynamic action to the registry.
+         *
+         * @param route The route instance to add
+         * @param action The corresponding instance of RestSendToExtensionAction to execute
+         */
+        public void registerDynamicRoute(RestHandler.Route route, RestSendToExtensionAction action) {
+            requireNonNull(route, "route is required");
+            requireNonNull(action, "action is required");
+            Optional<String> routeName = Optional.empty();
+            if (route instanceof NamedRoute) {
+                for (ActionType actionType : dynamicActionRegistry.registry.keySet()) {
+                }
+                routeName = Optional.of(((NamedRoute) route).name());
+                if (dynamicActionRegistry.isActionRegistered(routeName.get()) || registeredRestActionNames.contains(routeName.get())) {
+                    throw new IllegalArgumentException("route [" + route + "] already registered");
+                }
+            }
+            if (registry.containsKey(route)) {
+                throw new IllegalArgumentException("route [" + route + "] already registered");
+            }
+            registry.put(route, action);
+            routeName.ifPresent(registeredRestActionNames::add);
+        }
+
+        /**
+         * Remove a dynamic route from the registry.
+         *
+         * @param route The route to remove
+         */
+        public void unregisterDynamicRoute(RestHandler.Route route) {
```

Comment:
This would be useful for Hot Swap too: https://github.com/opensearch-project/opensearch-sdk-java/issues/356